dnsmasq.conf.j2

```diff
@@ -46,3 +46,13 @@ dhcp-option=3
 dhcp-option=6
 {% endif %}
 {% endif %}
+
+{%- if env["DHCP_ALLOWLIST"]|length %}
+{%- set dhcp_allowlist = env["DHCP_ALLOWLIST"].split(',') %}
+# An explicit list of hosts for which we should respond to DHCP requests:
+{%- for mac in dhcp_allowlist %}
+dhcp-host={{ mac }},net:allow
+{%- endfor %}
+dhcp-ignore=tag:!allow
+{%- endif %}
```

Comment:
We still have the problem that DHCP_RANGE could overlap - would it be better to specify an exact IP for each host (e.g pick the first N from DHCP_RANGE or something?

That would have the added benefit that the provisioning IPs for masters are predictable, and the reservations could be replicated in the cluster dnsmasq config?

We could do it with the ipmath filter similar to https://github.com/metal3-io/metal3-dev-env/blob/master/vm-setup/roles/libvirt/templates/network.xml.j2#L41 perhaps?

Reply:
MAO doesn’t know the macs of the masters now. How hard is that to get? @sadasu

Reply:
It could look a the BMH objects created by the installer, which contain the macs?

I guess the other option is not to care about the IPs being persistent, and just make sure the ranges don't overlap with some logic in the installer?

Reply:
So, I refactored this to take an env var with a semicolon separated list of `<MAC>,<IP>` leases.

I think the better solution is to give each master a static lease, *however* this is problematic in IPv6 because you can't use the MAC...you have to known the DUID... 

```
# Give a fixed IPv6 address and name to client with 
# DUID 00:01:00:01:16:d2:83:fc:92:d4:19:e2:d8:b2
# Note the MAC addresses CANNOT be used to identify DHCPv6 clients.
# Note also the they [] around the IPv6 address are obilgatory.
#dhcp-host=id:00:01:00:01:16:d2:83:fc:92:d4:19:e2:d8:b2,[1234::5] 
```

So, I guess maybe we *do* need to go with plan B and limit the DHCP range in the cluster to start from the n-th host.

Reply:
That wrinkle with IPv6 probably makes all of our plans not work.  It all really hinges on the bootstrap *not* responding to anything but the masters.

Reply:
TLDR: If we can't restrict the boostrap to only respond to the control plane, then a worker could *always* DHCP from the wrong DHCP server.  I think the only way to fix this is ensure that MAO absolutely does not start the metal3 pod until after bootstrap complete...

Reply:
I chatted with @stbenjam about this and I concur. We're trying to do IaaS, and filtering requests feels like we're trying to route around a race condition, when in reality, we should likely wait until our substrate is fully ready before we proceed to the next step. It means possibly a few more seconds or a minute depending on the configuration, but at least things become more reliable and we're not adding numerous conditionals and cases that have to be explored and tested each and every time.

Reply:
This is tricky, and I think there's enough nuance to solve this problem we should collect it all in one place. I wrote a doc describing the problem, this failed approach, and the way we might wait for MAO here:

https://docs.google.com/document/d/1uq7Kt1Ij8E0_Z-hYSCtV0loSep9hS01CUr0_ec316Pc/edit#

Please comment and edit as needed.

Reply:
Note that we have configured the DUID for RHCOS to be DUID-LL - this *is* derived from the MAC but it has a prefex, see here:

https://github.com/metal3-io/metal3-dev-env/pull/160/commits/500a37f392e5ea49462f2c66d1031ef5f1749fcd

https://github.com/openshift/machine-config-operator/pull/1395

The remaining part is @derekhiggins was looking into how we get IPA to send the same predictable ID (otherwise it gets a different IP and no hostname, in the case where there is a pool configured.

With this approach I think we should be able to configure dnsmasq for predictable leases, and I'd expect to be able to restrict things to only those reservations, but we'd have to test and confirm.

Reply:
But how do you boot into IPA then on real hardware? It looks like you’re setting it in libvirt for the dev-env, but a real metal host would need to DHCPv6 to boot into IPA and wouldn’t it just send the default way the host calculates the DUID?

Reply:
@stbenjam that's the part @derekhiggins has been looking into - the idea is to configure IPA the same way as we do for RHCOS so the DUID is the same in both cases (and thus the same DHCPv6 lease/reservation gets handed out)

Reply:
Right, I understand that, but I think I'm missing something fundamental here. To boot into IPA, you need to PXE off the DHCPv6 server first...

Are we *only* supporting virtual media on IPv6?

Reply:
@stbenjam with a specific client configuration (DUID-LL) the DUID is derived from the MAC address, similar to ipv4, so we know what it will be before the IPA boot happens.

This is different from the default (which is DUID-LLT I think), that is more designed for stateless ipv6 and has a timestamp component, thus it's impossible to predict.  That clearly isn't helpful when you want a stateful setup with reservations, hence configuring things (in IPA and RHCOS) to use DUID-LL instead.

Reply:
This PR should ensure that the IPA image has a predicable DUID
https://github.com/openshift/ironic-ipa-downloader/pull/26

Reply:
I updated this to use iptables instead, so we know it always works with direct clients in IPv6 and IPv4, without any worries about DUID's.